bin/varnishncsa/varnishncsa.c

```diff
@@ -163,8 +163,42 @@ static struct ctx {
 	int			recv_compl;
 } CTX;
 
+enum format_policy {
+        FMTPOL_FIRST = 0,
+        FMTPOL_REQ,
+        FMTPOL_RESP,
+};
+
 static void parse_format(const char *format);
 
+static unsigned
+process_format_policy(enum format_policy fp, unsigned first)
+{
+	unsigned policy;
+
+	if (fp == FMTPOL_FIRST)
+		return (first);
+
+#define BACKEND_FLAG 2<<8
+	policy = fp | ((*CTX.side == 'c') ? 0 : BACKEND_FLAG);
+
+	switch (policy) {
+		case FMTPOL_REQ:
+		case FMTPOL_RESP | BACKEND_FLAG:
+			if (CTX.recv_compl)
+				return (0);
+			return (first);
```

Comment:
In our offline discussion I hinted at decomposing the policy bits with boolean variables and returning a result in the end. Here we are still returning from everywhere and the new side-based macro is even more problematic.

We are nailing the semantics of our NCSA formats, it should be easy to understand the rules.